ui/src/VoxettaRecordButton.js

```diff
@@ -16,39 +16,85 @@ limitations under the License. */
 import {AudioRecorder} from './AudioRecorder';
 import {LitElement, html, css} from 'lit-element';
 
+/**
+ * Component responsible for recording and uploading audio files. 
+ */
 export class VoxettaRecordButton extends LitElement {
     static get properties() {
         return {
             isRecording: {type: Boolean},
             audioRecorder: {type: Object},
         };
     }
+
     constructor() {
         super();
         this.isRecording = false;
         this.audioRecorder = new AudioRecorder();
     }
-    render() {  
+
+    render() {
+        this.getUploadUrl();  
         return html`
             <button @click=${this.recordHandler}>Record Voice</button>
             <audio id="utterance" controls src="" style="display: none"></audio>
         `;
     }
 
-     async recordHandler(e){
+    /**
+     * If the user is not currently recording, begin recording using the Microphone component. Otherwise,
+     * stop recording and save and display the just-recorded audio file.
+     */
+    async recordHandler(){
         if (!this.isRecording) {
             this.isRecording = true;
             this.audioRecorder.startRecording();
         } else {
-            const audioSave = this.shadowRoot.getElementById("utterance");
             this.isRecording = false;
-            const recordingUrl = await this.audioRecorder.stopRecording();
-            if(recordingUrl != null){
-                audioSave.src = recordingUrl;
+            const audioSave = this.shadowRoot.getElementById("utterance");
+            const audio = await this.audioRecorder.stopRecording();
+            
+            // If non-empty, save and display the just-recorded audio file
+            if(audio.recordingUrl != null){
+                audioSave.src = audio.recordingUrl;
                 audioSave.style.display = "block";
+                this.saveAudio(audio);
             }
         }
     }
+
+    /**
+     * Set the blobUrl property to be a Blobstore upload link. 
+     */
+    async getUploadUrl() {
```

Comment:
This interaction with the servlets can be isolated into a stand-alone class like the `AudioRecorder` is. Maybe called `UtteranceApiService`? In this class you can handle. 

- Fetching the blob URL on creation
- Constructing Form
- Saving Audio

So this will give strong isolation of concerns. Record button is mostly focused on presentation, event handling and routing. AudioService is focused on interaction with browser audio APIs, and this new class is focused on network transmission, fetching data from and packing and sending data to the servlets.

Reply:
+1

Reply:
+1

Reply:
moved to UtteranceApiService